### PR TITLE
Add release and profile support for Windows

### LINKED ISSUE #38477: Implement release mode for Windows
Currently a release build (from the `flutter` tool perspective) is using a debug library build, in JIT mode; the only part that's different is anything at the Runner level (e.g., build flags for the native Runner code).

We should be building and uploading a release build of the engine library, and using that with AOT for [release mode builds](https://flutter.dev/docs/testing/build-modes), as on mobile.

### LINKED ISSUE #39664: Implement profile mode for Windows
Blocked on #38477. Once release mode is working, there are some tweaks necessary to the native build projects, and possibly to the tooling, to finish wiring up profile.

Some of the latter (e.g., adding profile mode to the VS project; it wouldn't do much for now, but really neither does release mode) could be done in advance.

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
## Description

Adds release and profile mode support to Windows.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/38477
Fixes https://github.com/flutter/flutter/issues/39664

## Tests

I added the following tests: Profile/release bundling copies app.so.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

Technically the now-removed test that non-debug builds output a warning about the lack of non-debug-build support failed, but Windows is explicitly unstable, and not yet covered by the breaking change policy.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by stuartmorgan** [comment]
I need to re-do the manual tests once the engine rolls so that I'm using actual artifact lookup rather than local engine artifact lookup, but it should be good for review.

File: packages/flutter_tools/lib/src/build_system/targets/windows.dart [hunk 1]
```diff
@@ -105,16 +105,13 @@ class DebugBundleWindowsAssets extends Target {
...
     Source.pattern('{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/windows.dart'),
     Source.pattern('{PROJECT_DIR}/pubspec.yaml'),
     ...IconTreeShaker.inputs,
   ];
 
   @override
-  List<Source> get outputs => const <Source>[
-    Source.pattern('{OUTPUT_DIR}/flutter_assets/kernel_blob.bin'),
-  ];
+  List<Source> get outputs => const <Source>[];
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/windows.dart:114]: Since this class is abstract anyway I would leave out the empty outputs

**Comment by stuartmorgan** [comment]
I wasn't entirely sure about inheritance vs. dependencies for the non-build-mode-specific parts of the Windows build. Let me know if I should restructure that base class into something that all the mode-specific targets depend on.

**Review by jonahwilliams — APPROVED** [review]
LGTM!

change seems reasonable overall. I can also test manually once the engine rolls.

File: packages/flutter_tools/lib/src/build_system/targets/windows.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/build_system/targets/windows.dart]: Moved down to the subclasses.

File: packages/flutter_tools/lib/src/build_system/targets/windows.dart [hunk 2]
```diff
@@ -149,3 +143,90 @@ class DebugBundleWindowsAssets extends Target {
     );
   }
 }
+
+/// A wrapper for AOT complitaion that copies app.so into the output directory.
```
Comment (MarcusTomlinson) [inline packages/flutter_tools/lib/src/build_system/targets/windows.dart:147]: complitaion -> compilation :)

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/build_system/targets/windows.dart]: Oops!

**Comment by stuartmorgan** [comment]
Manually tested debug, profile, and release (after locally merging in latest master, which has the needed roll), deleting the entire `build/windows/` directory each time just to be sure it wasn't relying on something that was bleeding across configurations, and all build and launch. This should be good to land once the tree is green.

**Review by jonahwilliams — APPROVED** [review]
Still LGTM

**Comment by Cretezy** [comment]
Question related to desktop release builds: Is it still using the full VM + snapshots or has x86 AOT compilation been added (doesn't seem like it but might've missed it)?

**Comment by texasbruce** [comment]
> 
> 
> Question related to desktop release builds: Is it still using the full VM + snapshots or has x86 AOT compilation been added (doesn't seem like it but might've missed it)?
> 
> Secondary question: what's the size for a base application on Windows?

Just tested it. In release build, it requires AOT to run. It is not copied directly to the data folder, which actually is required to run. You have to manually copy it (it is in the build/windows folder). Possible bug?

The minimum is around ~16M = AOT ~3M + flutter_windows.dll ~12M + assets ~1M

**Comment by stuartmorgan** [comment]
> It is not copied directly to the data folder, which actually is required to run. You have to manually copy it (it is in the build/windows folder). Possible bug?

You need to delete and recreate your `windows/` directory. The fact that the tool didn't enforce that was a bug whose fix just landed.

**Comment by stuartmorgan** [comment]
> has x86 AOT compilation been added

32-bit Windows continues to be unsupported. This was not intended to change that.

**Comment by texasbruce** [comment]
> 
> 
> > It is not copied directly to the data folder, which actually is required to run. You have to manually copy it (it is in the build/windows folder). Possible bug?
> 
> You need to delete and recreate your `windows/` directory. The fact that the tool didn't enforce that was a bug whose fix just landed.

I actually deleted the whole build folder, then did `build windows --release` again. 

app.so is under `build/windows` not `build/windows/x64/Release/Runner/data`.

**Comment by stuartmorgan** [comment]
> I actually deleted the whole build folder

My comment wasn't about the build folder. Please sync to the latest master, and you should get a failure with instruction when building.

**Comment by texasbruce** [comment]
> 
> 
> > I actually deleted the whole build folder
> 
> My comment wasn't about the build folder. Please sync to the latest master, and you should get a failure with instructions when building.

Sorry didn't realize there was more commits. Just did a pull and app.so is in data now. Looks good!

**Comment by Cretezy** [comment]
> > has x86 AOT compilation been added
> 
> 32-bit Windows continues to be unsupported. This was not intended to change that.

I think you misunderstood what I meant (sorry!). My question is that is it AOT like ARM, or is it still using the JIT on x86 platforms (which include x86_64)

**Pull request merged**